오픈마켓 II [STEP 2] Jiseong, Chacha

yim2627 opened the pull request:
안녕하세요 도미닉! @AppleCEO 

오픈마켓 2 Step2가 마무리되어 PR 보내드립니다.

전반적인 폴더 구조나, 역할과 책임을 고민하며 코드를 작성하였습니다!

이 부분에 대한 고민이 많았어서 관련하여 피드백주실 것이 있다면 말씀해주시면 감사하겠습니다! 😁

잘 부탁드립니다! 감사합니다~!

### 고민했던 점 

#### 1. 공통 기능 상속 구현 

상품 등록/수정 화면에서 요구되는 공통된 화면,기능들이 많았었기에, 이전 step에서는 프로토콜 기본 구현을 활용하여 공통 기능들을 구현해줬었습니다. 하지만 objective-c 메서드들은 프로토콜 기본 구현이 불가하기 때문에, 이번 step에서는 `상속`을 활용하여 공통 기능을 구현하고 이를 상속하여 기능들을 사용할 수 있도록 해줬습니다. 

각 하위 뷰 컨트롤러에서 가져야하는 특수한 화면이나, 기능들의 경우 상속 이후 자체적으로 내부에서 구현해줬습니다.

#### 2. 상품 수정/삭제 접근 권한 

상품 상세 화면에 진입하게 되면, 우측 상단에 상품 수정/삭제 버튼이 보여야 한다는 요구사항이 있었습니다. 

상품 수정/삭제의 경우 본인이 등록한 상품이 아닐 경우 수정/삭제가 불가능하기 때문에 초입부에서부터 접근을 막아주는 것이 효율적일 것이라 판단했습니다. 

이에 상품 상세 화면에 진입했을 때, 본인 상품인지 아닌지를 먼저 판별하여 화면에서의 버튼 노출 여부를 결정하였습니다. 

|본인이 등록 상품 상세 화면|타인이 등록한 상품 상세 화면|
|:---:|:---:|
|![](https://i.imgur.com/OP1NXVk.png)|![](https://i.imgur.com/SiEaFuX.png)|

#### 3. NotificationCenter를 활용한 업데이트 요청

상품 수정/삭제가 이루어지고 해당 화면이 내려가게 되면 이전 메인 화면이나 상품 상세 화면에서의 업데이트 또한 필요했습니다. 

이에 NotificationCenter의 userInfo에 업데이트된 데이터를 담아 최신 데이터로 업데이트 해줄 것을 요청하는 로직을 구현했습니다. 매번 업데이트를 요청하는 것이 아니라, 데이터의 변경사항에 발생했을 때만 업데이트를 보내주도록 하여 무분별한 API 요청을 하지 않도록 구현했습니다.

#### 4. 상세 화면 이미지를 화면에 꽉 차도록 리사이징

|Before Resize|After Resize|
|:--:|:--:|
|![](https://i.imgur.com/qbF7kyy.png)|![](https://i.imgur.com/cwPyPpd.png)|

`scaleAspectFill`을 쓰면 이미지가 확대되어 잘리는데, 저희가 원하는 모습은 이미지가 잘리지않는 모습이기때문에 이미지뷰의 속성을 `scaleAspectFit`을 사용하여 이미지가 잘리지 않게 구현하였습니다.

하지만 이미지가 잘리진 않아도, 양 옆에 하얀 공백이 생기는 것이 마음에 들지않아 이전에 구현해놓았던 `resize` 메서드를 활용하여 공백이 생기지 않도록 조정해주었습니다. 

### 궁금한 부분 

#### 1. 폴더 구조

파일이 많아지며 폴더의 구조를 변경시켜 개발자 본인이나, 다른 사람들이 보기 편하게끔 계층을 재구성하였습니다.

뷰와 뷰컨트롤러 부분은 Scene이라는 폴더로 위치시켜 구분이 되게하였습니다.

궁금한 점은 저희 코드 상 공통된 기능을 묶어 상위 뷰컨트롤러를 만들고, 해당 뷰컨트롤러를 상속받는 식으로 구현되어있는 뷰컨트롤러가 존재하는데, 상위 뷰컨트롤러를 어느 폴더에 속하게 만들어줘야할지가 고민되었습니다.

도미닉이 보시기엔 폴더 구조가 보기 편하신지 그리고 여러 폴더에 속해야하는 파일의 경우 어떤 방식으로 정리해주시는지 궁금합니다.

#### 2. 코드 구조 

계속 View와 View Controller의 책임을 나누고 코드도 나누어 작성하려고 노력했습니다!(말씀주신 viewController 100줄 이하는 다 지키지 못했지만...)

현재 저희 코드에서 코드 분리나 통합의 부분에서 개선 가능한 부분이 있을지 궁금합니다!
전체적으로 코드의 구조가 어떤지 궁금합니다!

yim2627 committed:
feat: ProductDetailScrollView 구현 #12

ChaminLee committed:
feat: 상품 상세 화면 구현 #12

- 상품 상세 화면 내 레이아웃 조정
- 상품 터치시 상품 상세 화면으로 이동하도록 구현

yim2627 committed:
feat: navigation 버튼 추가 및 버튼 액션 추가 #12

ChaminLee committed:
feat: 상품 이미지 슬라이더 구현 #12

- 상품 페이지 넘버 구현
- 메인 화면 리스트 레이아웃 조정
- 이미지 슬라이더 stackview에서 분리

yim2627 committed:
feat: 수정 api 요청 및 업데이트 기능 구현 #13

- 등록뷰의 이미지 스택뷰 속성 변경
- NotificationKey 타입 추가
- UserInformation 타입 추가
- 본인의 상품이 아닐 때 수정/삭제 불가능하도록 기능 구현

ChaminLee committed:
feat: 상품 삭제 기능 구현 #13

- 상품 secret 조회 기능 로직 수정(header추가, body 타입 변경)
- 상품 삭제 결과에 따른 alert 구현
- ProductSecret 타입 구현

yim2627 committed:
refactor: 전체 코드 접근제어자 및 구조 수정 #14

ChaminLee committed:
refactor: ProductManageViewController에 등록/수정 공통 기능 구현 #14

- 상품 등록/수정 ViewController는 ProductManageViewController 상속
- AlertMessage 추가
- productRegisterManager내 판별 로직 추가
- productManageable 제거

yim2627 committed:
chore: 파일 및 폴더 구조 변경 #14

- 화면별로 폴더 분리

yim2627 committed:
refactor: productDetailImage 잘리지 않게 리사이징 #14

ChaminLee committed:
feat: descriptionTextView placeholder 추가 #13

AppleCEO commented on OpenMarket/OpenMarket/Constants/AlertActionMessage.swift:
```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+enum AlertActionMessage {
```
이름이 적절하군요! 👍

AppleCEO commented on OpenMarket/OpenMarket/Constants/AlertMessage.swift:
```diff
@@ -0,0 +1,59 @@
+import Foundation
+
+enum AlertMessage: String {
```
이름이 조금 짧은 것 같네요.
AlertMessages 복수형은 어떨 것 같나요?

yim2627 replied:
타입의 경우 단수형으로 써주는 것이 더 좋다고 생각했습니다.

예를 들어 같은 열거형으로 정의하는 Error의 경우 내부 에러 케이스가 여러가지라고 해서 Errors라고 하지않는 것처럼 위 코드도 동일하게 해야되지 않나 라는 생각이 들었는데, 도미닉의 의견은 어떠신가요? 😁😁

AppleCEO commented on OpenMarket/OpenMarket/Constants/UserInformation.swift:
```diff
@@ -0,0 +1,5 @@
+enum UserInformation {
```
이 이름도 짧아보이네요!
조금 더 명확한 이름 있을까요?

yim2627 replied:
`VendorInformation` 으로 수정해주었습니다!

AppleCEO commented on OpenMarket/OpenMarket/Models/Request/ProductSecret.swift:
```diff
@@ -0,0 +1,9 @@
+import Foundation
+
+struct ProductSecret: Codable {
+    let secret: String
```
ProductSecret 은 상품 비밀이라고 해석되는데요.
어떤 역할의 구조체인지 이름으로 알기에는 난해한 것 같네요.
ProductSecret.secret 이라는 프로퍼티도 어떤 역할인지 불분명하네요.

ChaminLee replied:
서버 단에서 secret이라는 네이밍을 사용해서 그대로 차용했었는데요, `ProductPassword`가 더 명확할 것 같아 변경하였습니다!

AppleCEO commented on OpenMarket/OpenMarket/Scene/Detail/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,217 @@
+final class ProductDetailViewController: UIViewController {
+    private let productDetailScrollView = ProductDetailScrollView()
+    private var productDetail: ProductDetail
+    private let api = APIService()
```
api 라는 프로퍼티명도 좀 짧은 것 같습니다.
API 관련 코드들을 모아놓은 것은 APIService 객체로 생성한 것 같은데 apiService 라는 이름도 괜찮을 것 같군요!

yim2627 replied:
apiService가 타입명을 정확히 표현해주는 것 같아 `apiService`로 수정해주었습니다. 감사합니다!

AppleCEO commented on OpenMarket/OpenMarket/Scene/Detail/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,217 @@
+    }
+    
+    @objc private func receiveModifiedProductDetail(_ notification: Notification) {
+        if let modifiedProductDetail = notification.userInfo?[NotificationKey.detail] as? ProductDetail {
+            fetchProductDetail(with: modifiedProductDetail)
+            productDetail = modifiedProductDetail
+        }
```
if let 을 사용하셨군요~
저는 이렇게 한번 정도 사용되는 변수를 옵셔널 바인딩할 때는 if let 을 선호합니다!

AppleCEO commented on OpenMarket/OpenMarket/Scene/Detail/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,217 @@
+            return
+        }
+        
+        let imageWidth: CGFloat = view.frame.width
```
여기에 CGFloat 타입을 명시적으로 선언해준 이유가 있을까요?

ChaminLee replied:
타입 자체가 `CGFloat`이기에 명시적으로 표기하지 않아도 될 것 같아 제거하였습니다!

AppleCEO commented on OpenMarket/OpenMarket/Scene/Detail/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,217 @@
+            let imageView = UIImageView()
+            let xPosition = imageWidth * CGFloat(index)
+            let image = ImageLoader.loadImage(from: images[index].url)
+            let resizedImage = image?.resizeImageTo(size: CGSize(width: imageWidth, height: ProductDetailImageSize.height))
```
상수명들이 명확하네요! 💯

AppleCEO commented on OpenMarket/OpenMarket/Scene/Detail/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,217 @@
+        }
+        let cancelAction = UIAlertAction(title: AlertActionMessage.cancel.title, style: .cancel, handler: nil)
+        
+        [modifyAction, deleteAction, cancelAction].forEach { action in
+            alert.addAction(action)
```
기본형을 활용한 부분 좋습니다~ 🥇

AppleCEO commented on OpenMarket/OpenMarket/Scene/Detail/Controller/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,217 @@
+// MARK: - Delete Product
+
+extension ProductDetailViewController {
+    private func checkSecret(secret: String?) {
```
비밀을 확인한다는 메소드도 조금 불명확한 것 같아요! 🔢

ChaminLee replied:
메서드를 `CheckProductPassword(from password: )`로 변경해주었습니다!

또한 해당 메서드의 역할을 분리해주고자 하여 completion handler를 추가하여 비밀번호를 성공적으로 조회하면 반환하도록 했습니다. 이에 이 메서드의 클로저에서 성공/실패 케이스에 따라 상품 삭제 성공/싪패에 따른 액션을 수행하도록 변경하였습니다! 

추가적으로 상품을 삭제하는 과정이 completion을 연달아 계속 타면서 실행이 되는 것 같은데, 이에 흐름을 파악하기가 조금 어렵다는 생각이 들었습니다. 현재 구조의 문제인지, 역할을 나눔에 따라 자연스레 발생하는 것인지 궁금합니다!

AppleCEO replied:
check라는 메소드명은 별로 활용하지 않는게 좋다고 하더라고요!
저도 주로 사용했었는데 무엇을 체크한다는건지, 체크하고 결과가 어떨지 좀 불명확한 면이 있었습니다.
is 로 시작하게 메소드명을 지으면 리턴값이 Bool 값이겠군요! 라고 더 명확하게 표현할 수 있는 것 같아요.
이 부분 참고해서 다시 메소드명 지어봐도 좋겠습니다~

ChaminLee replied:
반환값이 Bool 타입이 아니어서 `is`는 좋지 않을 것 같다고 생각했습니다! 말씀주신 것 처럼 Check가 애매한 것 같아서 `fetchProductPassword`라고 지어주었습니다. 비밀번호를 가져오는 것이 주된 역할이기 때문에 이렇게 네이밍 해보고자 합니다! (가져오는데 있어서 성공/실패의 케이스는 있지만 불러온다는 fetch가 포함할 수 있지 않을까 생각했습니다)

AppleCEO commented on OpenMarket/OpenMarket/Scene/Register/View/CustomSegmentedControl.swift:
```diff
@@ -1,6 +1,6 @@
 import UIKit
 
-class LayoutSegmentedControl: UISegmentedControl {
+class CustomSegmentedControl: UISegmentedControl {
```
커스텀이라는 이름보다 역할을 나타내는 단어를 붙이는 것이 더 좋습니다!
세그먼트 컨트롤을 상속받아 다르게 만들면 다 커스텀이죠~
어떤 역할을 하는 세그멘트 컨트롤인지 나타낼 수 있는 이름으로 개선해보면 좋겠습니다.

ChaminLee replied:
해당 Segmented Control자체가 메인 화면에서의 list/grid 전환, KRW/USD 전환에 사용되고 있어 공통된 네이밍을 찾기에는 어려움이 있어 보였습니다. 이에 `CustomSegmentedControl`이라는 네이밍을 붙였는데요! 여러 군데에서 사용되고 있다보니 특정 역할을 담아주기 어려운 것 같은데.. 이 경우에도 네이밍 변경이 필요할까요??

AppleCEO replied:
ㅎㅎ 그랬군요! 그럼 프로젝트 이름을 넣는 경우도 종종 있었어요.
OpenMarketSegement 같은 이름은 커스텀보다는 조금 더 성의있는 이름이 될 수 있을 것 같아요~

ChaminLee replied:
`OpenMarketSegement` 으로 변경해보았습니다!

AppleCEO commented on OpenMarket/OpenMarket/Scene/Modify/ProductModifyViewController.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+    
+    @objc private func didTapDoneButton() {
+        if !checkValidInput() {
```
유효한 입력을 확인한다라고 변역되는데 조금 불명확한 것 같습니다. 이 경우 더 명확한 메소드 이름이 있을까요?

yim2627 replied:
`checkValidReceivedProductDetail` 로 네이밍을 변경하여 어떠한 입력을 검증하는지 명확하게 표현해주었습니다!!

AppleCEO replied:
네 조금 더 명확해졌네요.
바뀐 이름에서 check 를 is 로 바꾸면 어떨까요?

yim2627 replied:
Bool 타입의 값을 반환하는 함수이므로 말씀해주신 `isValidReceiveProductDetail`이 더 좋은 것 같다는 생각이 들어 그렇게 바꿨습니다!

AppleCEO commented on OpenMarket/OpenMarket/Scene/Register/Controller/ProductRegisterManager.swift:
```diff
@@ -15,7 +15,7 @@ class ProductRegisterManager {
     
     @objc private func didTapAddImageButton() {
-        delegate?.presentImagePickerView()
+        pickerPresenterDelegate?.presentImagePickerView()
```
이름이 더 길면서 명확해졌네요~ 👍

AppleCEO commented on OpenMarket/OpenMarket/Scene/Register/Controller/ProductRegisterManager.swift:
```diff
@@ -108,16 +127,20 @@ class ProductRegisterManager {
     func fetchRegisteredProductDetail(from productDetail: ProductDetail) {
-        guard let image = ImageLoader.loadImage(from: productDetail.thumbnail) else {
-            return
+        if let productImages = productDetail.images {
+            productImages.forEach { productImage in
+                guard let image = ImageLoader.loadImage(from: productImage.url) else {
+                    return
+                }
+                addImageToImageStackView(from: image, hasDeleteButton: false)
+            }
```
흔하지는 않은 구조 같네요.
if let 안에 guard let이 있군요!
이 부분 조금 더 알아보기 좋게 개선해 볼 수 있을까요?

ChaminLee replied:
```swift
guard let productImages = productDetail.images else {
    return
}

productImages.forEach { productImage in
    guard let image = ImageLoader.loadImage(from: productImage.url) else {
        return
    }
    addImageToImageStackView(from: image, hasDeleteButton: false)
}
```

위 같은 스타일로 변경해주고자 합니다!

AppleCEO replied:
오호 뎁스도 줄어들고 괜찮아보이네요~

AppleCEO reviewed (changes requested):
안녕하세요 도미닉입니다
@ChaminLee & @yim2627 고생하셨습니다.
속도가 굉장히 빨랐군요!

질문에 대한 답변
1. 폴더 구조
- 독특한 구조네요! 저는 보통 뷰컨트롤러끼리 모아주고 그 중에서 공통되는 특징으로 또 폴더에 모아주고 하는 식으로 진행합니다. 예를 들면 ViewController/Scrap 이렇게 내려가는 구조에요. 이렇게 역할에 따라 먼저 나눠주는 것은 신선하네요~ 자신과 팀원에게 맞는 구조를 여러 방법으로 시도해보고 맞춰가보는 식으로 많이 진행합니다. 이런 방식으로 한번 체계화 해보는 것도 좋은 경험일 것 같네요.

2. 코드 구조
- 메인 뷰컨트롤러 파일에 enum 들이 있군요! 보통 파일 하나에 하나의 객체, 프로토콜, 열거형만을 저장합니다. 이렇게 되면 뷰컨트롤러가 더 길어보이기도 하고 열거형을 찾을 때도 찾기 힘들 수 있겠네요~ 개선해보고 가능하다면 전에 이야기했던 UICollectionViewDelegate 등을 MainViewController 가 아닌 다른 객체에게 할당해보는 것도 진행해보면 좋겠네요.

yim2627 committed:
refactor: UserInformation -> VendorInformation 네이밍 수정 #14

ChaminLee committed:
refactor: ProductSecret -> ProductPassword 네이밍 변경 #14

yim2627 committed:
refactor: api -> apiService 프로퍼티 네이밍 변경 #14

ChaminLee committed:
refactor: 불필요한 타입 명시 제거 #14

yim2627 committed:
refactor: 상품 삭제 및 비밀번호 조회 로직 변경 #14

- checkValidInput -> checkValidReceivedProductDetail 네이밍 변경

ChaminLee committed:
refactor: 조건문 들여쓰기 개선 #14

yim2627 committed:
feat: refreshControl 추가 #15

ChaminLee committed:
feat: 새 상품 업로드 유무를 확인하고, 업데이트하기 위한 버튼 구현 #15

- 버튼을 내리기 위해 애니메이션 구현

yim2627 committed:
refactor: 타입, 메서드의 네이밍 변경 #14

q
qa!

ChaminLee committed:
feat: List화면 Pagination 구현 #15

- LoadingCollectionViewReusableView 구현
- 다음 페이지가 존재할 경우 새로운 데이터 추가
- CollectionView List Layout 구현 방식 변경

ChaminLee committed:
refactor: ImageLoad 비동기로 변경 및 prefetch 구현 #14

- collectionView 스크롤 버벅임 해결
- prefetch에서 다음 페이지의 데이터를 불러오도록 구현
- cell 높이 고정적으로 변경

AppleCEO commented:
고생하셨습니다. viewWillAppear 함수를 적극적으로 활용한 부분 좋았습니다.
이런 기본적으로 제공되는 함수들을 요긴하게 사용하면 원하는 시점에 동작을 잘 시킬 수 있어서 좋습니다!
이야기할 부분은 다 한 것 같아서 머지 원하시면 말씀해주세요~

ChaminLee committed:
docs: Step4 Readme 작성

ChaminLee committed:
docs: Step 4 Readme 수정

ChaminLee committed:
docs: Step 4 Readme 최종 수정

ChaminLee committed:
docs: readme 수정
